crypto/pkcs7/pk7_doit.c

```diff
@@ -55,6 +55,43 @@ ASN1_OCTET_STRING *PKCS7_get_octet_string(PKCS7 *p7)
     return NULL;
 }
 
+static ASN1_OCTET_STRING *pkcs7_get1_data(PKCS7 *p7)
+{
+    ASN1_OCTET_STRING *os = PKCS7_get_octet_string(p7);
+
+    if (os != NULL) {
+        /* Edge case for MIME content, see RFC 5652 section-5.2.1 */
+        ASN1_OCTET_STRING *osdup = ASN1_OCTET_STRING_dup(os);
+
+        if (osdup != NULL && (os->flags & ASN1_STRING_FLAG_NDEF))
+            /* ASN1_STRING_FLAG_NDEF flag is currently used by openssl-smime */
+            ASN1_STRING_set0(osdup, NULL, 0);
```

Comment:
Can you comment on why clearing the data is necessary?  Is it just to avoid looking at the NDEF flag later, because empty outputs are handled just like NDEF outputs? Or is there something else non-obvious going on?

Reply:
It simplifies calling ASN1_OCTET_STRING_free(), see https://github.com/openssl/openssl/pull/22575#discussion_r1675602857

Reply:
Removing the line `ASN1_STRING_set0(osdup, NULL, 0);` causes the following memory leak when running the` openssl smime -sign` command:
```
==84352== HEAP SUMMARY:
==84352==     in use at exit: 1 bytes in 1 blocks
==84352==   total heap usage: 9,915 allocs, 9,914 frees, 775,183 bytes allocated
==84352== 
==84352== 1 bytes in 1 blocks are definitely lost in loss record 1 of 1
==84352==    at 0x484280F: malloc (vg_replace_malloc.c:442)
==84352==    by 0x4C0D105: CRYPTO_malloc (mem.c:202)
==84352==    by 0x4C0D270: CRYPTO_realloc (mem.c:295)
==84352==    by 0x4A91377: ASN1_STRING_set (asn1_lib.c:314)
==84352==    by 0x4A911EB: ASN1_STRING_copy (asn1_lib.c:262)
==84352==    by 0x4A91276: ASN1_STRING_dup (asn1_lib.c:279)
==84352==    by 0x4A889D3: ASN1_OCTET_STRING_dup (a_octet.c:16)
==84352==    by 0x4CFC96F: pkcs7_get1_data (pk7_doit.c:64)
==84352==    by 0x4CFD1CC: PKCS7_dataInit (pk7_doit.c:280)
==84352==    by 0x4CFB950: pk7_cb (pk7_asn1.c:46)
==84352==    by 0x4A97103: BIO_new_NDEF (bio_ndef.c:99)
==84352==    by 0x4A92D3F: i2d_ASN1_bio_stream (asn_mime.c:77)
==84352== 
==84352== LEAK SUMMARY:
==84352==    definitely lost: 1 bytes in 1 blocks
==84352==    indirectly lost: 0 bytes in 0 blocks
==84352==      possibly lost: 0 bytes in 0 blocks
==84352==    still reachable: 0 bytes in 0 blocks
==84352==         suppressed: 0 bytes in 0 blocks
==84352== 
==84352== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 0 from 0)

```

Reply:
> Removing the line `ASN1_STRING_set0(osdup, NULL, 0);` causes the following memory leak when running the` openssl smime -sign` command:
> 
> ```
> ==84352== HEAP SUMMARY:
> ==84352==     in use at exit: 1 bytes in 1 blocks
> ==84352==   total heap usage: 9,915 allocs, 9,914 frees, 775,183 bytes allocated
> ==84352== 
> ==84352== 1 bytes in 1 blocks are definitely lost in loss record 1 of 1
> ==84352==    at 0x484280F: malloc (vg_replace_malloc.c:442)
> ==84352==    by 0x4C0D105: CRYPTO_malloc (mem.c:202)
> ==84352==    by 0x4C0D270: CRYPTO_realloc (mem.c:295)
> ==84352==    by 0x4A91377: ASN1_STRING_set (asn1_lib.c:314)
> ==84352==    by 0x4A911EB: ASN1_STRING_copy (asn1_lib.c:262)
> ==84352==    by 0x4A91276: ASN1_STRING_dup (asn1_lib.c:279)
> ==84352==    by 0x4A889D3: ASN1_OCTET_STRING_dup (a_octet.c:16)
> ==84352==    by 0x4CFC96F: pkcs7_get1_data (pk7_doit.c:64)
> ==84352==    by 0x4CFD1CC: PKCS7_dataInit (pk7_doit.c:280)
> ==84352==    by 0x4CFB950: pk7_cb (pk7_asn1.c:46)
> ==84352==    by 0x4A97103: BIO_new_NDEF (bio_ndef.c:99)
> ==84352==    by 0x4A92D3F: i2d_ASN1_bio_stream (asn_mime.c:77)
> ==84352== 
> ==84352== LEAK SUMMARY:
> ==84352==    definitely lost: 1 bytes in 1 blocks
> ==84352==    indirectly lost: 0 bytes in 0 blocks
> ==84352==      possibly lost: 0 bytes in 0 blocks
> ==84352==    still reachable: 0 bytes in 0 blocks
> ==84352==         suppressed: 0 bytes in 0 blocks
> ==84352== 
> ==84352== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 0 from 0)
> ```

Interesting! This could be argued to be a bug in `ASN1_STRING_copy`, since it copies the flags (including NDEF), and should know to not copy any data, in that case, since it won't be freed?  Or did I misunderstand something here?

Reply:
@olszomal What do you think? Is the need to explicitly zero out the duplicated string to avoid a leak a bug in how NDEF strings are duplicated?

Reply:
I'd say it could be seen as a bug but it should be resolved in a subsequent PR.

Reply:
Agree with @t8m